### PR TITLE
refactor(devtools): add types to Explorer

### PR DESCRIPTION
**Pull request by Liam-Tait**
Add types to Explorer component with as minimal functional changes as
possible while still getting type safety

#2742

### CODE REVIEW

**Comment by TkDodo** [comment]
can you try adding some more tests please

**Comment by Liam-Tait** [comment]
> can you try adding some more tests please

Can do

**Comment by Liam-Tait** [comment]
@TkDodo are you be happy for me to do some additional clean up in this file too?

**Comment by TkDodo** [comment]
> @TkDodo are you be happy for me to do some additional clean up in this file too?

sure :)

**Comment by TkDodo** [comment]
@Liam-Tait this looks like a nice addition, but the PR is still in `draft` state. Is ready for review? Thank you

**Comment by Liam-Tait** [comment]
@TkDodo I think this still needed some more tests. I haven't added enough to maintain or increase coverage, however most of these changes are adding types to existing code.

File: src/devtools/Explorer.tsx [hunk 1]
```diff
@@ -45,43 +52,81 @@ export const Expander = ({ expanded, style = {}, ...rest }) => (
...
+type RendererProps = {
+  HandleEntry: HandleEntryComponent
+  label?: string
+  value?: any
+  subEntries?: Entry[]
+  subEntryPages?: Entry[][]
+  type?: string
+  expanded?: boolean
+  toggleExpanded: () => void
+  pageSize: number
```
Comment (TkDodo) [inline src/devtools/Explorer.tsx:68]: it looks like we could make most (all?) of these props mandatory, as we always pass them in (there is only one usage)

File: src/devtools/Explorer.tsx [hunk 2]
```diff
@@ -45,43 +52,81 @@ export const Expander = ({ expanded, style = {}, ...rest }) => (
...
-const DefaultRenderer = ({
-  handleEntry,
+type Entry = {
+  label: string
+}
+
+type RendererProps = {
+  HandleEntry: HandleEntryComponent
+  label?: string
+  value?: any
```
Comment (TkDodo) [inline src/devtools/Explorer.tsx:62]: can we use `unknown` here instead of `any`?

**Review by TkDodo — COMMENTED** [review]
I'm fine with the test coverage. You've added types AND added a test. not sure how the coverage can go down but that doesn't matter :)

File: src/devtools/Explorer.tsx (hunk 2 above)
Comment (Liam-Tait) [reply src/devtools/Explorer.tsx]: Added, I needed to add an extra check that the value isIterable

5bc51d8949ec0be5f6e6cf9811f344ec198ff6f6

File: src/devtools/Explorer.tsx (hunk 1 above)
Comment (Liam-Tait) [reply src/devtools/Explorer.tsx]: Remove optional for all except `label` 3b4aeef4bfa5ed8b3838593cc9558b39785bb4fc

**Comment by Liam-Tait** [comment]
@TkDodo Addressed issues, ready for a re-review

**Review by TkDodo — APPROVED** [review]

**Comment by TkDodo** [comment]
seems like there are some type errors now:

```
Error: src/devtools/tests/Explorer.test.tsx(39,11): error TS2322: Type '{ label: string; toggleExpanded: Mock<any, any>; pageSize: number; handleEntry: () => Element; expanded: false; subEntryPages: { label: string; }[][]; }' is not assignable to type 'IntrinsicAttributes & RendererProps'.
``

**Comment by Liam-Tait** [comment]
@TkDodo Fixed

**Comment by TkDodo** [comment]
@Liam-Tait I'm now seeing some react key related warnings when running the react-query devtools tests:

``
console.error
    Warning: Each child in a list should have a unique "key" prop.
    
    Check the render method of `Explorer`. See https://reactjs.org/link/warning-keys for more information.
        at HandleEntry (react-query/src/devtools/Explorer.tsx:242:21)
        at Explorer (react-query/src/devtools/Explorer.tsx:185:3)
```

can you maybe have a look at those please?

**Comment by TkDodo** [comment]
@Liam-Tait we sadly have 2 more (minor) issues reported that are likely introduced with this PR:

---

1) `Data Explorer` now shows white background for expandable fields:

3.4.16:

Screenshot 2022-04-06 at 21 25 24

3.4.17:
Screenshot 2022-04-06 at 21 24 08

was this an intended change?

---

2) expanded entries in the explorer now collapse if there is a background refetch that didn't actually change anything.

For example, take a look at [this sandbox](https://codesandbox.io/s/expand-devtools-90jt1x)
- click on the query in the devtools, and expand the `owner` field
- then, re-focus the window

--> the `owner` collapses.

If you take the sandbox back to `3.4.16`, `owner` stay expanded.

---

Could you take a look at these two things please? Should we open separate issues? Thank you!

**Comment by TkDodo** [comment]
ah, issue 1 is caused by wrapping the clickable element in a `button`, but it's unstyled.

https://github.com/tannerlinsley/react-query/pull/2949/files#diff-a29eb5ef62b495218e2308396eb9064e59c6630c92421a6ceef78a0836d07fe3R17

**Pull request merged**